src/polynomial.rs

```diff
@@ -1,343 +1,471 @@
+
+    pub fn iter_mut(&mut self) -> IterMut<F> {
+        self.0.iter_mut()
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.0.iter().all(|x| x.is_zero())
+    }
+
+    /// Number of coefficients held by the polynomial. Is NOT equal to the degree in general.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Degree of the polynomial.
+    pub fn degree(&self) -> usize {
+        (0usize..self.len())
+            .rev()
+            .find(|&i| self[i].is_nonzero())
+            .expect("Zero polynomial")
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Removes the coefficients in the range `range`.
+    fn drain<R: RangeBounds<usize>>(&mut self, range: R) {
+        self.0.drain(range);
+    }
+
+    /// Evaluates the polynomial at a point `x`.
+    pub fn eval(&self, x: F) -> F {
+        self.iter().rev().fold(F::ZERO, |acc, &c| acc * x + c)
+    }
+
+    /// Evaluates the polynomial at a point `x`, given the list of powers of `x`.
+    /// Assumes that `self.len() == x_pow.len()`.
+    pub fn eval_from_power(&self, x_pow: &[F]) -> F {
+        F::inner_product(&self[..], x_pow)
+    }
+
+    /// Evaluates the polynomial on subgroup of `F^*` with a given FFT precomputation.
+    pub fn eval_domain(&self, fft_precomputation: &FftPrecomputation<F>) -> Vec<F> {
+        let domain_size = fft_precomputation.size();
+        if self.len() < domain_size {
+            // Need to pad the polynomial to have the same length as the domain.
+            fft_with_precomputation(&self.padded(domain_size).coeffs(), fft_precomputation)
+        } else {
+            fft_with_precomputation(&self.coeffs(), fft_precomputation)
         }
-        (quotient, remainder)
     }
-}
 
-fn inv_mod_xn<F: Field>(h: &Polynomial<F>, n: usize) -> Polynomial<F> {
-    assert!(h[0].is_nonzero(), "Inverse doesn't exist.");
-    let mut hh = h.clone();
-    if h.len() < n {
-        for _ in 0..n - h.len() {
-            hh.push(F::ZERO);
+    /// Computes the interpolating polynomial of a list of `values` on a subgroup of `F^*`.
+    pub fn from_evaluations(values: &[F], fft_precomputation: &FftPrecomputation<F>) -> Self {
+        Self(ifft_with_precomputation_power_of_2(
+            values,
+            fft_precomputation,
+        ))
+    }
+
+    /// Leading coefficient.
+    pub fn lead(&self) -> F {
+        self.iter()
+            .rev()
+            .find(|x| x.is_nonzero())
+            .map_or(F::ZERO, |x| *x)
+    }
+
+    /// Reverse the order of the coefficients, not taking into account the leading zero coefficients.
+    fn rev(&self) -> Self {
+        let d = self.degree();
+        Self(self.0[..=d].iter().rev().copied().collect())
+    }
+
+    /// Negates the polynomial's coefficients.
+    fn neg(&self) -> Self {
+        Self(self.iter().map(|&x| -x).collect())
+    }
+
+    /// Removes leading zero coefficients.
+    pub fn trim(&mut self) {
+        if !self.is_zero() {
```

Comment:
Maybe nice to have a `degree_plus_1` method (in addition to `degree`), for cases like this where calling `degree` could fail